src/sciline/domain.py

```diff
@@ -1,10 +1,11 @@
 # SPDX-License-Identifier: BSD-3-Clause
 # Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
-from typing import Any, Generic, TypeVar
+from typing import Generic, TypeVar
 
-T = TypeVar("T")
+PARAM = TypeVar("PARAM")
+SUPER = TypeVar("SUPER")
 
 
-class Scope(Generic[T]):
-    def __new__(cls, x) -> Any:  # type: ignore[no-untyped-def]
+class Scope(Generic[PARAM, SUPER]):
```

Comment:
> Using SUPER type-var is to bind the x: SUPER to the specific super type right...?

Yes, otherwise `mypy` will not detect if you create the type with an incompatible value.

Reply:
> From perspective of the static type checker and dependency injection,
> it's not necessary to inherit `SUPER` like
> `class Float(Scope[PARAM, float], float): ... `
> and we can just do `class Float(Scope[PARAM, float]): ...` instead.

This is not correct: This *is* needed for type-checking, otherwise you get errors about missing methods or operators that were not found. It is not required for *runtime*.

Reply:
> Is there any sort of way to detect if the other super class matches as the second type argument ...?
> For example,
> `class Float(Scope[PARAM, float], str): ...` should raise an error either during the type-checking or runtime somehow I thought.

I wanted to do that too, but I could not think of an easy way. So if you cannot either... :(

Reply:
Right ... I missed it cause I was not using any specific method of the class... now I understood...!

Reply:
> I wanted to do that too, but I could not think of an easy way. So if you cannot either... :(

The only one I could think of was something like this
```
class Scope(Generic[PARAM, SUPER]):

    def __init_subclass__(cls) -> None:
        supertype = get_args(cls.__orig_bases__[0])[1]
        if supertype not in cls.__bases__:
            raise TypeError(f"{cls} should inherit {supertype}.")
        return super().__init_subclass__()

    def __new__(cls, x: SUPER) -> SUPER:  # type: ignore[misc]
        return x
```

I just tried this naively and it seems work....?

Reply:
I had to do a couple more changes, but it seems to work so far.